### PR TITLE
FIX correct node splitting order & remove class weight 

### PR DESCRIPTION
**Pull request by PSSF23**
#### Reference Issues/PRs

https://github.com/neurodata/scikit-tree/issues/107

### CODE REVIEW

File: sklearn/ensemble/_forest.py
Comment (adam2392) [inline sklearn/ensemble/_forest.py:1]: Are the CI errors related possibly to these changes?

Comment (PSSF23) [reply sklearn/ensemble/_forest.py:1]: I don't see why other checks all succeeded. Rerunning the failed one now...

**Review by PSSF23 — COMMENTED** [review]
The error still persists at forest level. I'll test more about it. Not ready to merge yet.

**Comment by PSSF23** [comment]
@adam2392 When you have time can you try running the code several times? The error is randomly appearing on my machine, and I couldn't find any clue why yet. Thanks.

```python
from sklearn.ensemble import RandomForestClassifier

from sklearn import datasets
from numpy.random import permutation

iris = datasets.load_iris()

iris_X = iris.data
iris_y = iris.target
p = permutation(iris_X.shape[0])
iris_X = iris_X[p]
iris_y = iris_y[p]

clf_1 = RandomForestClassifier()
clf_1.partial_fit(iris_X, iris_y, classes=[0,1,2])
clf_1.partial_fit(iris_X, iris_y)
```

**Comment by PSSF23** [comment]
Didn't find anything suspicious about `tree.pyx` that could cause the memory crash. Maybe the problem lies elsewhere? Like `Splitter`?

**Comment by adam2392** [comment]
> Didn't find anything suspicious about `tree.pyx` that could cause the memory crash. Maybe the problem lies elsewhere? Like `Splitter`?

Hmmm... I'm not sure, but this error now looks like some of the errors I was getting very randomly in scikit-tree, which I just skipped this specific test. I couldn't spot any errors at the time, but if it's coming up like this, then it means it could be an error I presume.

Some thoughts:

- versions of joblib? I presume because it's related to some parallelization over trees
- some edge-case of the partial fit, you are accessing memory that is not allocated, or deallocated. This could be happening in the update_node, or update_tree part in Cython... If so, this is worrying.
- Some version of the compiler? I don't see why though.

I can't reproduce the error on my machine tho forsure.

**Review by PSSF23 — COMMENTED** [review]
@adam2392 I think I found the problem. `build` was resizing `tree` each time it runs, but that step seems to disrupt update. After correcting it, I don't see the error anymore.

See b92c9eefe

**Comment by adam2392** [comment]
Hmmm dam that would be a finicky bug. Is there any unit test we can add that checks this at the tree level that fails on main but succeeds in this PR branch?

**Review by PSSF23 — COMMENTED** [review]
There appears to be other CI errors that I didn't resolve. Working on them now.

For unit tests, the problem is no test can handle segmentation fault? Unless there is one I don't know.

**Comment by adam2392** [comment]
> There appears to be other CI errors that I didn't resolve. Working on them now.
> 
> For unit tests, the problem is no test can handle segmentation fault? Unless there is one I don't know.

As in the unit-test results in segmentation fault reliably in main, but does not in this branch.

Is it true that the root cause is that the capacity changes upon `partial_fit` twice? If that is the case, would a valid test be assert the capacity does not change (assuming the new data you feed in does not increase the depth of the tree significantly).

**Comment by PSSF23** [comment]
I'll keeping looking into the errors then. The resizing fix definitely resolved my local errors (code above).

**Comment by adam2392** [comment]
> I'll keeping looking into the errors then. The resizing fix definitely resolved my local errors (code above).

I cannot replicate this error locally, but I'm inclined to merge this for now. If you're able to investigate on your end, then that would be great. I believe the error points to some type of fault in the partial_fit implementation since the traceback consistently shows it errors out in the `self.builder_.build(...)` line, which means there is some error in the Cython code when building the tree via partial fit for this test.

**Pull request merged**